platform/include/retry_utils.h

```diff
@@ -31,6 +31,72 @@
+ * The retry utilities are a set of APIs aiding in retrying with exponential 
+ * backoff and jitter. Exponential backoff with jitter is strongly recommended 
+ * for retrying failed actions over the network with servers. Please see 
+ * https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/ for 
+ * more information about the benefits with AWS.
+ * 
+ * Exponential backoff with jitter is typically used when retrying a failed
+ * connection to the server. In an environment with poor connectivity, a client
+ * can get disconnected at any time. A backoff strategy helps the client to
+ * conserve battery by not attempting reconnections over and over when it is
+ * unlikely to succeed.
+ * 
+ * Before retrying the failed communication to the server there is a quiet period.
+ * In this quiet period, the task that is retrying must sleep for some random 
+ * amount of time between 0 and a base and a maximum number of seconds. The base is
+ * doubled with each retry attempt until the maximum is reached.<br>
+ * 
+ * > sleep_seconds = random_between( 0, min( 2<sup>attempts_count</sup> * base_seconds, maximum_seconds ) )
+ * 
+ * @section retryutils_implementation Implementing Retry Utils
+ * 
+ * The functions that must be implemented are:<br>
+ * - @ref RetryUtils_ParamsReset
+ * - @ref RetryUtils_BackoffAndSleep
+ * 
+ * The functions are used as shown in the diagram below. This is the exponential
+ * backoff with jitter loop:
+ * 
+ * @image html retry_utils_flow.png
+ * 
+ * The following steps give guidance on implementing the Retry Utils:
+ * 
+ * -# Implementing @ref RetryUtils_ParamsReset
+ * @snippet this define_retryutils_paramsreset
+ *<br>
+ * This function initializes @ref RetryUtilsParams_t. It is expected to set
+ * @ref RetryUtilsParams_t.attemptsDone to zero. It is also expected to set 
+ * @ref RetryUtilsParams_t.nextJitterMax to @ref INITIAL_RETRY_BACKOFF_SECONDS 
+ * plus some random amount of seconds, jitter. This jitter is a random number 
+ * between 0 and @ref MAX_JITTER_VALUE_SECONDS. This function must be called before 
+ * entering the exponential backoff with jitter loop using 
+ * @ref RetryUtils_BackoffAndSleep.<br><br>
+ * 
+ * -# Implementing @ref RetryUtils_BackoffAndSleep
+ * @snippet this define_retryutils_backoffandsleep
+ * <br>
+ * When this function is invoked, the calling task is expected to sleep a random
+ * number of seconds between 0 and @ref RetryUtilsParams_t.nextJitterMax. After 
+ * sleeping this function must double @ref RetryUtilsParams_t.nextJitterMax, but
+ * not exceeding @ref MAX_RETRY_BACKOFF_SECONDS. When @ref MAX_RETRY_ATTEMPTS are
+ * reached this function should return @ref RetryUtilsRetriesExhausted, unless
+ * @ref MAX_RETRY_ATTEMPTS is set to zero.
+ * When @ref RetryUtilsRetriesExhausted is returned the calling application can
+ * stop trying with a failure, or it can call @ref RetryUtils_ParamsReset again
+ * and restart the exponential back off with jitter loop.
```

Comment:
So I was looking at our implementation of the retry utilities and most of it is platform independent, or can be rewritten to be so (i.e. replacing clock_gettime with srand(time(NULL))). The one exception is `sleep()`. I wonder if it'd be easier to just paste our code and say "replace this with your platform's sleep function"

Reply:
I feel like the extra works of adding these instructions and even adding FIXMEs for others would have been avoided if each API asked for callbacks to get time and sleep.